src/main/java/org/spongepowered/common/SpongeLifecycle.java

```diff
@@ -199,6 +202,20 @@ public void callLoadedGameEvent() {
     public void callStoppingEngineEvent(final Engine engine) {
         this.game.eventManager().post(SpongeEventFactory.createStoppingEngineEvent(PhaseTracker.getCauseStackManager().currentCause(),
                 engine, this.game, (TypeToken<Engine>) TypeToken.get(engine.getClass())));
+        if (engine instanceof SpongeServer) {
+            final @Nullable SpongeGameProfileManager profileManager = ((SpongeServer) engine).gameProfileManagerIfPresent();
+            if (profileManager != null) {
+                profileManager.close();
+            }
+        }
+    }
+
+    public void callStoppingGameEvent() {
+        // Call an event for plugins to shut down any thread pools
+        this.game.eventManager().post(SpongeEventFactory.createStoppedGameEvent(PhaseTracker.getCauseStackManager().currentCause(), this.game));
+
+        // Then shut down our own thread pool
+        ((AsyncScheduler) this.game.asyncScheduler()).close();
```

Comment:
Would it be better to do this in a shutdown hook?

Reply:
If we did that, the async scheduler's thread pool would have to use deamon threads, which loses us control, and the ability to warn on stuck tasks when the server is shutting down.

Reply:
Ohhh I see now, did not consider that.